resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/AutoBuildJobRescheduleTest.java

```diff
@@ -0,0 +1,200 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Stefan Winkler and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Stefan Winkler - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.core.tests.internal.builders;
+
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.eclipse.core.resources.*;
+import org.eclipse.core.runtime.*;
+import org.eclipse.core.runtime.jobs.*;
+import org.eclipse.core.tests.harness.TestBarrier2;
+import org.eclipse.core.tests.internal.builders.TestBuilder.BuilderRuleCallback;
+
+/**
+ * Test for #158
+ * (https://github.com/eclipse-platform/eclipse.platform/issues/158).
+ *
+ * The basic problem case of this bug seems to be a race condition in which a
+ * thread modifies the workspace and requests a build while the AutoBuildJob is
+ * running.
+ *
+ * To test this, we use a somewhat illegal {@link ISchedulingRule} that makes it
+ * possible to actually modify the workspace while the build is running. We are
+ * using the {@link TestBarrier2} mechanism to synchonize the two concurrent
+ * threads to get a deterministic flow of events.
+ */
+public class AutoBuildJobRescheduleTest extends AbstractBuilderTest {
+	// state constants to be used for the barrier
+	private static final int AUTOBUILD_RUNNING_BLOCKED = 1;
+	private static final int AUTOBUILD_CONTINUE = 2;
+	private static final int AUTOBUILD_DONE = 3;
+
+	/**
+	 * The barrier to synchronize the builder and the concurrent workspace
+	 * modification
+	 */
+	private TestBarrier2 barrier = new TestBarrier2();
+
+	/**
+	 * A file in the workspace that we modify to trigger builds
+	 */
+	private IFile file;
+
+	/**
+	 * A thread-safe flag to record the rescheduling of the AutoBuildJob
+	 */
+	private AtomicBoolean rescheduled = new AtomicBoolean(false);
+
+	/**
+	 * A somewhat illegal {@link ISchedulingRule} which satisfies the sanity checks
```

Comment:
If this is illegal as per contract, we won't add support for it, but instead `JobManager.validateRule` should be improved if possible to reject it.

Reply:
I am not fixing broken scheduling rules. The original issue is created by a race condition in a scenario that uses correct scheduling rules. I was asked to write a test case to reproduce the issue, but the nature of a race condition is that you can’t reproduce it. So I found this hole in the scheduling rule check which makes it possible to have a test case.